### PR TITLE
Allow round-tripping ASCII masked tables for most formats

### LINKED ISSUE #5337: Cannot round-trip masked value to ECSV
Astropy 1.2.1, Python 3.5.2, macOS 10.11.

A colleague reported being unable to load an ECSV file with a masked value into a masked Astropy table. Here is code that reproduces the exception:

``` python
from astropy import table
a = [1, 4, 5]
b = [2.0, 5.0, 8.2]
c = ['x', 'y', 'z']
t = table.Table([a, b, c], names=('a', 'b', 'c'), masked=True, dtype=('i4', 'f8', 'S1'))
t['a'].mask = [True, True, False]
t.write('./temp.ecsv', format='ascii.ecsv')

table.Table.read('./temp.ecsv', format='ascii.ecsv')
```

The table file:

```
# %ECSV 0.9
# ---
# datatype:
# - {name: a, datatype: int32}
# - {name: b, datatype: float64}
# - {name: c, datatype: string}
a b c
-- 2.0 x
-- 5.0 y
5 8.2 z
```

The exception:

```
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
/.../lib/python3.5/site-packages/astropy/io/ascii/core.py in _convert_vals(self, cols)
    942                 try:
--> 943                     converter_func, converter_type = col.converters[0]
    944                     if not issubclass(converter_type, col.type):

IndexError: list index out of range

During handling of the above exception, another exception occurred:

ValueError                                Traceback (most recent call last)
<ipython-input-5-d4804d245b66> in <module>()
      8 t.write('./temp.ecsv', format='ascii.ecsv')
      9
---> 10 table.Table.read('./temp.ecsv', format='ascii.ecsv')

/.../lib/python3.5/site-packages/astropy/table/table.py in read(cls, *args, **kwargs)
   2330         passed through to the underlying data reader (e.g. `~astropy.io.ascii.read`).
   2331         """
-> 2332         return io_registry.read(cls, *args, **kwargs)
   2333
   2334     def write(self, *args, **kwargs):

/.../lib/python3.5/site-packages/astropy/io/registry.py in read(cls, *args, **kwargs)
    349
    350         reader = get_reader(format, cls)
--> 351         data = reader(*args, **kwargs)
    352
    353         if not isinstance(data, cls):

/.../lib/python3.5/site-packages/astropy/io/ascii/connect.py in io_read(format, filename, **kwargs)
     35     from .ui import read
     36     format = re.sub(r'^ascii\.', '', format)
---> 37     return read(filename, format=format, **kwargs)
     38
     39

/.../lib/python3.5/site-packages/astropy/io/ascii/ui.py in read(table, guess, **kwargs)
    339                                              ' with fast (no guessing)'})
    340         else:
--> 341             dat = reader.read(table)
    342             _read_trace.append({'kwargs': new_kwargs,
    343                                 'status': 'Success with specified Reader class '

/.../lib/python3.5/site-packages/astropy/io/ascii/core.py in read(self, table)
   1178
   1179         self.data.masks(cols)
-> 1180         table = self.outputter(cols, self.meta)
   1181         if hasattr(self.header, 'table_meta'):
   1182             table.meta.update(self.header.table_meta)

/.../lib/python3.5/site-packages/astropy/io/ascii/core.py in __call__(self, cols, meta)
    970
    971     def __call__(self, cols, meta):
--> 972         self._convert_vals(cols)
    973
    974         # If there are any values that were filled and tagged with a mask bit then this

/.../lib/python3.5/site-packages/astropy/io/ascii/core.py in _convert_vals(self, cols)
    957                     last_err = err
    958                 except IndexError:
--> 959                     raise ValueError('Column {} failed to convert: {}'.format(col.name, last_err))
    960
    961

ValueError: Column a failed to convert: invalid literal for int() with base 10: '--'
```

Anonymized user comments:

> I'm wondering if it because the first line is masked, because I realised I have another file with a masked value and reads in ok. So in that case, indeed maybe a bug. Still, in the interest of doing science and not arsing around with error messages: bogus values for the win at the moment!
> 
> That's rather vexing, I agree, since masking entries is the Right Way To Do It and this seems to be pushing you towards using bogus sentinel values like in the `-99.99 mag` days
> 
> I have no experience with the ecsv file format as such but have found masked values a bit tricky to work with in astropy…. I usually end up using bogus values too.

### PR DESCRIPTION
**Pull request by taldcroft**
This PR allows round-tripping ASCII masked tables for most formats by changing the default `io.ascii` masked value output from `--` to an empty string.

This is a relatively significant API change, but backward compatible output is possible by adding `fill_values=[(ascii.masked, '--')]` to any existing `write` statements that require the old convention.

Overall I think the API break is worth it given the substantial benefit.  It turns out that certain things in the existing output `fill_values` machinery were not working or inconsistent anyway.  For instance the slow writers use the numpy default string for a masked value (`np.ma.masked_print_option`) while the fast writers always use `--`.

As for the actual table value representing "missing data", the precedent of an empty string from CSV seems compelling and all formats support this already.  Otherwise one needs to make up some arbitrary sentinel value since `--` is not unique enough.  The fact that the readers _already_ interpret an empty value as missing in every format is good and means there is no API change from that perspective.

Closes #5337.

The current version does not pass tests for  ['fixed_width', 'aastex', 'latex'].   `fixed_width` needs some special handling in the test.  The `tex` formats should work pending the tex lastline fix (e.g. #5236).

Closes #5235.
This PR includes complete round-tripping tests.

cc: @hamogu @astrofrog

### CODE REVIEW

File: astropy/io/ascii/core.py [hunk 1]
```diff
@@ -1427,6 +1427,13 @@ def _get_writer(Writer, fast_writer, **kwargs):
 
     from .fastbasic import FastBasic
 
+    # A values of None for fill_values imply getting the default string
+    # representation of masked values (depending on the writer class), but the
+    # machinery expects a list.  The easiest here is to just pop the value off,
+    # i.e. fill_values=None is the same as not providing it at all.
+    if 'fill_values' in kwargs and kwargs['fill_values'] is None:
```
Comment (MSeifert04) [inline astropy/io/ascii/core.py]: This seems overly verbose, isn't this equivalent to `kwargs.get('fill_values', None) is None`

File: astropy/io/ascii/core.py [hunk 2]
```diff
@@ -1463,7 +1470,8 @@ def _get_writer(Writer, fast_writer, **kwargs):
     if 'exclude_names' in kwargs:
         writer.exclude_names = kwargs['exclude_names']
     if 'fill_values' in kwargs:
-        writer.data.fill_values = kwargs['fill_values']
+        # Prepend user-specified values to the class default.  Both are lists.
+        writer.data.fill_values = kwargs['fill_values'] + writer.data.fill_values
```
Comment (MSeifert04) [inline astropy/io/ascii/core.py]: This breaks some tests because one is a tuple and the other a list. Not sure why (user/test-input?) but maybe this needs a `tuple` or `list` call around it (or `itertools.chain` and a cast afterwards).

**Comment by MSeifert04** [comment]
Am I right, that with this change some or most of the files previously written may not be re-read correctly?

File: astropy/io/ascii/core.py (hunk 1 above)
Comment (MSeifert04) [inline astropy/io/ascii/core.py]: and maybe `del kwargs['fill_values']` is more explicit than `pop`.

Comment (taldcroft) [inline astropy/io/ascii/core.py]: Agreed on the `del kwargs`.

The `kwargs.get()` expression is not the same if `fill_values` is not in the supplied `kwargs`, which is normally the case.

**Comment by taldcroft** [comment]
@MSeifert04 - This change has no impact on reading previously written files since the reader is not changing.  What it will disrupt is the case of an existing code that is writing new files for a masked table and then relying on the output having the `--` fill values for masked data.  For instance the code might be reading back using the `fill_values=[('--', '0')]` trick.

So of course we need to have a serious discussion about whether the breakage is going to be widely problematic.  The release notes would have some fairly strong notes about this.

File: astropy/io/ascii/core.py (hunk 2 above)
Comment (taldcroft) [inline astropy/io/ascii/core.py]: According to the docs `fill_values` for `write()` needs to be a list of tuples, but there has been a convenience built in to the reader-side that if there was only one tuple then you don't need the list around it.  It looks like that short-cut is built into the tests.  Urgh.

File: astropy/io/ascii/core.py (hunk 1 above)
Comment (MSeifert04) [inline astropy/io/ascii/core.py]: you're right. I was thinking about the `dict.pop` with default value, in that case `kwargs.get` would've worked - but with `del` you're approach is the only one that makes sense.

**Comment by MSeifert04** [comment]
@taldcroft Just a guess but the `intp` (default int-type) for windows is `int32` (even for 64bit builds). I assume if you have trouble round-tripping that somewhere `dtype=int` is used and in another place `int64` (or some equivalent). Might be somewhere in the fast-reader code.

**Comment by taldcroft** [comment]
@MSeifert04 - yes, it's definitely a problem related to `int32` being default on Windows.  #4686 was my failed attempt to fix this.  I was hoping that the fast reader would recognize an int32 overflow and upcast (19670dc) but this didn't fly and in fact made a bunch of other stuff break.  In the meantime I'm just going to a slightly weaker test and only require that the dtype kind matches so `i4` and `i8` will be considered equal.  There is no strict requirement of matching at this level for ASCII formats that don't include detailed data type information.

**Comment by taldcroft** [comment]
Rebased off of #5370 which should fix the Windows failures.

**Comment by embray** [comment]
I feel like this should probably do something with `DEFAULT_ASCII_TNULL` in the FITS module while we're at it, but I'm not sure exactly what (except maybe change its value to `''`): https://github.com/astropy/astropy/blob/master/astropy/io/fits/column.py#L131

**Comment by taldcroft** [comment]
@embray - I would not discount the idea of adjusting `DEFAULT_ASCII_TNULL`, but I'm just a bit more scared of messing around with FITS output since I imagine that this might have a larger impact owing to the wider use of FITS in "production" code vs simple ASCII tables.  I have no real evidence for this assertion of course...

In any case I would prefer to keep that idea separate from this PR.

File: astropy/io/ascii/core.py (hunk 2 above)
Comment (hamogu) [inline astropy/io/ascii/core.py]: Well, we wanted to test that the short-cut works.

File: astropy/io/ascii/core.py [hunk 3]
```diff
@@ -671,7 +671,7 @@ class BaseData(object):
     fill_include_names = None
     fill_exclude_names = None
     # Currently, the default matches the numpy default for masked values.
```
Comment (hamogu) [inline astropy/io/ascii/core.py]: If you change the default, this comment line has to go. Note however, that this also states why we chose "--" as default in the first place - it's consistent with numpy. Maybe astropy is mature enough that we don't need that any longer, if we want the "least surprising" behavior, doing what numpy does is not a bad choice.

**Comment by taldcroft** [comment]
BTW this clashes with #5379 and should probably wait till that is merged.

File: astropy/io/ascii/core.py (hunk 3 above)
Comment (taldcroft) [inline astropy/io/ascii/core.py]: Point taken about the least surprising behavior, but I think here we are getting into the distinction between the screen representation of numpy vs. a file serialization format.  I think it is pretty clear that `'--'` is no good as a unique serialization marker for missing data, whereas `''` has a fair bit of precedent.  And since we already took the much bigger  leap of making `''` indicated masked when *reading*, let's close the loop now and do the same for *writing*.

Comment (hamogu) [inline astropy/io/ascii/core.py]: I'm not opposed to this change, I just wanted to make sure it's a conscious decision to move away from the numpy format since that was the reason for having -- in the first place.

**Comment by taldcroft** [comment]
@hamogu - this is ready for final review.  The last 5 commits are new since you last looked.  There are no real functional changes, just docs and cleaner code and more complete testing.

**Comment by taldcroft** [comment]
Closes #5235.
This PR includes complete round-tripping tests.

File: astropy/io/ascii/core.py [hunk 4]
```diff
@@ -671,8 +671,7 @@ class BaseData(object):
     write_spacer_lines = ['ASCII_TABLE_WRITE_SPACER_LINE']
     fill_include_names = None
     fill_exclude_names = None
-    # Currently, the default matches the numpy default for masked values.
-    fill_values = [(masked, '--')]
+    fill_values = [(masked, '')]
```
Comment (hamogu) [inline astropy/io/ascii/core.py]: I would add a comment in here: `# Note that this differs from the numpy default. See #5347`

File: astropy/io/ascii/core.py [hunk 5]
```diff
@@ -1430,6 +1429,13 @@ def _get_writer(Writer, fast_writer, **kwargs):
 
     from .fastbasic import FastBasic
 
+    # A value of None for fill_values imply getting the default string
```
Comment (hamogu) [inline astropy/io/ascii/core.py]: imply -> implies

File: astropy/io/ascii/core.py [hunk 6]
```diff
@@ -1466,7 +1472,13 @@ def _get_writer(Writer, fast_writer, **kwargs):
...
         writer.exclude_names = kwargs['exclude_names']
     if 'fill_values' in kwargs:
-        writer.data.fill_values = kwargs['fill_values']
+        # Prepend user-specified values to the class default.
+        with suppress(TypeError, IndexError):
+            # Test if it looks like (match, replace_string, optional_colname),
+            # in which case make it a list
+            kwargs['fill_values'][1] + ''
+            kwargs['fill_values'] = [kwargs['fill_values']]
+        writer.data.fill_values = kwargs['fill_values'] + writer.data.fill_values
```
Comment (hamogu) [inline astropy/io/ascii/core.py]: This is (potentially) a change in behaviour. Previously, user defined fill values would replace writer defaults, now they are added to writer defaults. I believe that that's not a big issue for our built-in classes, because we only have defaults for None, missing, empty etc. However, I have build classes for my own use that derive from BaseReader and replace "-999" with missing. However, my collaborators sometimes also send me tables where missing value are marked with other numbers. In that case I used to call my own class with `fill_value='+999'`. In the new implementation, this would not treat both +999 and -999 as missing. For my particular case all data value are between -10 and +10, so I'm fine either way, I just wanted to give an example where is might actually matter. 

To summarize: I think this change (very few people build their own reader classes) is OK, but should be mentioned in the change log.

File: astropy/io/ascii/tests/test_write.py [hunk 7]
```diff
@@ -629,3 +639,41 @@ def test_write_overwrite_ascii(format, fast_writer, tmpdir):
...
+    Round trip a simple masked table through every writable format and confirm
+    that reading back gives the same result.
+    """
+    fmt_name, fmt_cls = fmt_name_class
+
+    if not getattr(fmt_cls, '_io_registry_can_write', True):
+        return
+
+    # Skip tests for fixed_width or HTML without bs4
+    if ((fmt_name == 'html' and not HAS_BEAUTIFUL_SOUP)
```
Comment (hamogu) [inline astropy/io/ascii/tests/test_write.py]: Pytest offers a short-cut for this case of "skip if import not successful":
``bs4 = pytest.importorskip("bs4")``

File: astropy/io/ascii/core.py (hunk 6 above)
Comment (taldcroft) [inline astropy/io/ascii/core.py]: @hamogu - I'm getting a little confused about whether you talking about reading or writing, since you mentioned getting files from your collaborators and replacing -999 with missing.  For writing you would have said "replacing missing with -999".

It's important to note that the user-supplied kwarg will always override the default by virtue of being earlier in the list.  I also think there will be no change in behavior for custom writer classes that define a custom `fill_value`, e.g. to output -999 for missing values.

Comment (hamogu) [inline astropy/io/ascii/core.py]: Sorry. You are right, I was getting confused, please disregard my example. However, the point still stands.
If I had a custom writer class with the following replacements:
missing -> -999
nan -> -888
'abc'->'def'
and I call it with `fill_values = ('123', '456')` then before this PR I would only have the 123->456 replacement. After this PR I will have 123->456 *plus* the default replacements missin->-999, nan->-888, and 'abs'->'cde'. Right?

In other words, before this change, I could switch off the default fill values by passing in essentially an empty list (or non-sense replacements that never happen. Now, I would have ot know the default replacements. If I want to witch them off in the example above, I would have to do (pseudo code) `fill_values=[(missing, missing), (nan,nan), ('abc', 'abc')]`.
Usually, default replacements in a class make sense and you usually want to keep them even if the user passes in a fill_value specificer to replace some other type of values, but it's stiull a change in behaviour.

File: astropy/io/ascii/tests/test_write.py (hunk 7 above)
Comment (taldcroft) [inline astropy/io/ascii/tests/test_write.py]: Cool.  In the interest of not further clogging CI, I would like to defer this change for now.  Anyway it should probably be done as a global change across astropy if we decide this is now the preferred idiom.  See #5543.

**Comment by taldcroft** [comment]
@hamogu - updated the changelog in 3f00763.  Look OK?

**Comment by wmwv** [comment]
@taldcroft   Awesome!  Thanks for fixing things up here and in #5427 and adding the round-tripping test.
